nntrainer/engine.cpp

```diff
@@ -0,0 +1,181 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2024 Jijoong Moon <jijoong.moon@samsung.com>
+ *
+ * @file   engine.cpp
+ * @date   27 December 2024
+ * @brief  This file contains engine context related functions and classes that
+ * manages the engines (NPU, GPU, CPU) of the current environment
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Jijoong Moon <jijoong.moon@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ */
+#include <dirent.h>
+#include <dlfcn.h>
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#include <app_context.h>
+#include <base_properties.h>
+#include <context.h>
+#include <engine.h>
+
+#ifdef ENABLE_OPENCL
+#include <cl_context.h>
+#endif
+
+static std::string solib_suffix = ".so";
+static std::string contextlib_suffix = "context.so";
```

Comment:
Maybe dlopen OS wrapper should handle solib_suffix; It's no-go for Windows with '.so' exension.

Such function would require only filename stem adding platform specific extension '.dll' vs '.so' vs '.dylib' if we'll ever make MacOS port.